docs/docs/typelevel.md

```diff
@@ -0,0 +1,383 @@
+returns optionally the default value of `T`, if it exists. In fact, we can already express
+this using ordinary match expressions and a simple helper function, `scala.typelevel.anyValue`, which is defined as follows:
+```scala
+def anyValue[T]: T = ???
+```
+The `anyValue` function _pretends_ to return a value of its type argument `T`. In fact, it will always raise a `NotImplementedError` exception instead.
+This is OK, since the function is intended to be used only on the type-level; it should never be executed at run-time.
+
+Using `anyValue`, we can then define `defaultValue` as follows:
+```scala
+transparent def defaultValue[T]: Option[T] = anyValue[T] match {
+  case _: Byte => Some(0: Byte)
+  case _: Char => Some(0: Char)
+  case _: Short => Some(0: Short)
+  case _: Int => Some(0)
+  case _: Long => Some(0L)
+  case _: Float => Some(0.0f)
+  case _: Double => Some(0.0d)
+  case _: Boolean => Some(false)
+  case _: Unit => Some(())
+  case _: t >: Null => Some(null)
+  case _ => None
+}
+```
+Then:
+```scala
+defaultValue[Int] = Some(0)
+defaultValue[Boolean] = Some(false)
+defaultValue[String | Null] = Some(null)
+defaultValue[AnyVal] = None
+```
+
+As another example, consider the type-level inverse of `toNat`: given a _type_ representing a Peano number, return the integer _value_ corresponding to it. Here's how this can be defined:
+```scala
+transparent def toInt[N <: Nat]: Int = anyValue[N] match {
+  case _: Z => 0
+  case _: S[n] => toInt[n] + 1
+}
+```
+
+## Computing New Types
+
+The examples so far all computed _terms_ that have interesting new types. Since in Scala terms can contain types it is equally easy to compute the types of these terms directly as well. To this purpose, it is helpful
+to base oneself on the helper class `scala.typelevel.Typed`, defined as follows:
+```scala
+class Typed[T](val value: T) { type Type = T }
+```
+Here is a version of `concat` that computes at the same time a result and its type:
+```scala
+transparent def concatTyped(xs: Tuple, ys: Tuple): Typed[_ <: Tuple] = xs match {
+  case ()       => Typed(ys)
+  case (x, xs1) => Typed((x, concatTyped(xs1, ys).value))
+}
+```
+
+## Avoiding Code Explosion
+
+Recursive transparent functions implement a form of loop unrolling through inlining. This can lead to very large generated expressions. The code explosion can be avoided by calling typed versions of the transparent functions to define erased values, of which  just the typed part is used afterwards. Here is how this can be done for `concat`:
+
+```scala
+def concatImpl(xs: Tuple, ys: Tuple): Tuple = xs match {
+  case ()       => ys
+  case (x, xs1) => (x, concatImpl(xs1, ys))
+}
+
+transparent def concat(xs: Tuple, ys: Tuple): Tuple = {
+  erased val r = concatTyped(xs, ys)
+  concatImpl(xs, ys).asInstanceOf[r.Type]
+}
```

Comment:
`erased` here seems to suggest that we need something like __transparent types__ as experimented in #4727 , which can express computations on types.